File: server/src/main/java/multiplayer/networking/messages/MessageType.java
package multiplayer.networking.messages;

public enum MessageType {
    // Incoming messages
    PLAYER_SHOOT_MESSAGE_FROM_CLIENT("playerShootMessageFromClient"),
    PLAYER_MOVE_FROM_CLIENT("playerMoveFromClient"),
    PLAYER_MOUSE_DIRECTION_FROM_CLIENT("playerMouseDirectionFromClient"),

    // Outgoing messages
    GAME_STATE("gameState"),
    HIT("hit"),
    PLAYER_DEATH("playerDeath"),
    WELCOME("welcome"),
    PLAYER_LEFT("playerLeft"),
    PLAYER_JOINED("playerJoined"),
    INITIAL_GAME_STATE("initialGameState"),
    GAME_STATE_SYNC("gameStateSync"),
    PLAYER_MOVE_FROM_SERVER("playerMoveFromServer");

    private final String type;

    MessageType(String type) {
        this.type = type;
    }

    public String getType() {
        return type;
    }

    public static MessageType fromString(String type) {
        for (MessageType messageType : MessageType.values()) {
            if (messageType.type.equals(type)) {
                return messageType;
            }
        }
        throw new IllegalArgumentException("Unknown message type: " + type);
    }
}

File: client/rust/src/network/messages/message_type.rs
use std::collections::HashMap;

use godot::prelude::GodotClass;
use serde::{Deserialize, Serialize};

use crate::utils::serializable_vector2::SerializableVector2;

use super::{bullet_data::BulletData, player_data::PlayerData};

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
#[serde(tag = "type")] // Use the "type" field in JSON to determine the variant.
pub enum MessageType {
    #[serde(rename_all = "camelCase")]
    GameStateSync {
        players: HashMap<String, PlayerData>,
        bullets: HashMap<String, BulletData>,
    },
    #[serde(rename_all = "camelCase")]
    Welcome {
        player_id: String,
        player_data: PlayerData,
    },

    #[serde(rename_all = "camelCase")]
    PlayerLeft { player_data: PlayerData },
    #[serde(rename_all = "camelCase")]
    PlayerJoined { player_data: PlayerData },
    #[serde(rename_all = "camelCase")]
    PlayerShootMessageFromClient { player_id: String },
    #[serde(rename_all = "camelCase")]
    PlayerMoveFromClient {
        player_id: String,
        direction: SerializableVector2,
        move_message_type: String,
    },
    #[serde(rename_all = "camelCase")]
    PlayerMouseDirectionFromClient {
        player_id: String,
        mouse_direction: SerializableVector2,
    },
    #[serde(rename_all = "camelCase")]
    SaveGameState,
    #[serde(rename_all = "camelCase")]
    LoadGameState,
}

impl MessageType {
    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }

    pub fn to_json(message: &MessageType) -> Result<String, serde_json::Error> {
        serde_json::to_string(message)
    }
}
